crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -411,42 +476,176 @@ impl Object {
-        let version = version.map(|v| v as i64);
+        use checkpoints::dsl as checkpoints;
+        use objects_history::dsl as history;
+        use objects_snapshot::dsl as snapshot;
 
-        let stored_obj: Option<StoredObject> = db
+        let results: Option<Vec<StoredHistoryObject>> = db
             .execute(move |conn| {
-                conn.first(move || {
-                    let mut query = dsl::objects
-                        .filter(dsl::object_id.eq(address.clone()))
+                conn.results(move || {
+                    // If an object was created or mutated in a checkpoint outside the current
+                    // available range, and never touched again, it will not show up in the
+                    // objects_history table. Thus, we always need to check the objects_snapshot
+                    // table as well.
+                    let mut snapshot_query = snapshot::objects_snapshot
+                        .filter(snapshot::object_id.eq(vec_address.clone()))
+                        .into_boxed();
+
+                    let mut historical_query = history::objects_history
+                        .filter(history::object_id.eq(vec_address.clone()))
+                        .order_by(history::object_version.desc())
                         .limit(1)
                         .into_boxed();
 
-                    // TODO: leverage objects_history
                     if let Some(version) = version {
-                        query = query.filter(dsl::object_version.eq(version));
+                        snapshot_query =
+                            snapshot_query.filter(snapshot::object_version.eq(version));
+                        historical_query =
+                            historical_query.filter(history::object_version.eq(version));
                     }
 
-                    query
+                    let left = snapshot::objects_snapshot
+                        .select(snapshot::checkpoint_sequence_number)
+                        .order(snapshot::checkpoint_sequence_number.desc())
+                        .limit(1);
+
+                    // if some checkpoint_sequence_number is provided, use it as the upper bound.
+                    // This is because an object can be extant at some target, but be indexed with
+                    // at a checkpoint_sequence_number <= target.
+                    if let Some(checkpoint_sequence_number) = checkpoint_sequence_number {
+                        // A validation check is not needed as we coalesce both sides to 0 for null
+                        // values, and if left > right, the db will return 0 rows. This also saves
+                        // us a roundtrip.
+                        historical_query = historical_query.filter(
+                            history::checkpoint_sequence_number
+                                .nullable()
+                                .between(left.single_value(), checkpoint_sequence_number),
```

Comment:
between left is inclusive, we only need history objects after that, so this should be exclusive?

Reply:
Ah yeah that's fair, left would be from objects_snapshot, so can be exclusive on objects_history